saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java

```diff
@@ -371,13 +371,14 @@ public static Converter<ResponseToken, Saml2ResponseValidatorResult> createDefau
 			Response response = responseToken.getResponse();
 			Saml2AuthenticationToken token = responseToken.getToken();
 			Saml2ResponseValidatorResult result = Saml2ResponseValidatorResult.success();
-			String statusCode = getStatusCode(response);
-			if (!StatusCode.SUCCESS.equals(statusCode)) {
-				String message = String.format("Invalid status [%s] for SAML response [%s]", statusCode,
-						response.getID());
-				result = result.concat(new Saml2Error(Saml2ErrorCodes.INVALID_RESPONSE, message));
+			List<String> statusCodes = getStatusCodes(response);
```

Comment:
Instead of getting all the status codes ad infinitum, please just get the top-level one. 

If it is of type `REQUESTER`, `RESPONDER`, or `VERSION_MISMATCH`, then add a single error message that includes this status code and any single child status code.

Otherwise, if it isn't `SUCCESS`, then add a single error message like it already does.

This is based on the logic outlined in: https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf#page=39